lib/gc/Dialect/Microkernel/MicrokernelOps.cpp

```diff
@@ -8,7 +8,369 @@
+    return op.emitOpError() << "expect ldb to be >= of dimension n\n";
+  if (ldc < n)
+    return op.emitOpError() << "expect ldc to be >= of dimension n\n";
+
+  // Verify dispatch flags.
+  return verifyBrgemmFlags(op.getFlags(), op, FLAGS_NAME);
+}
+
+/////////////////////////////////////////////////////
+// Start of BrgemmOp
+
+// TODO(haixin): could use compiler-wide VNNI utils?
+static bool isInVnniLayout(MemRefType memref) {
+  if (!memref.getElementType().isBF16() &&
+      !memref.getElementType().isSignedInteger(8) &&
+      !memref.getElementType().isUnsignedInteger(8)) {
+    return false;
+  }
+
+  auto blockingFactor = 0;
+  if (memref.getElementType().isBF16()) {
+    blockingFactor = 2;
+  } else if (memref.getElementType().isSignedInteger(8) ||
+             memref.getElementType().isUnsignedInteger(8)) {
+    blockingFactor = 4;
+  }
+  return memref.getShape().back() == blockingFactor;
+}
+
+static bool isTypeCompatible(Type outType, Type operandAType,
+                             Type operandBType) {
+  if (!outType.isF32() && !outType.isSignedInteger(32)) {
+    return false;
+  }
+  if (outType.isF32()) {
+    if (!(operandAType.isF32() && operandBType.isF32()) &&
+        !(operandAType.isBF16() && operandBType.isBF16())) {
+      return false;
+    }
+  }
+  if (outType.isSignedInteger(32)) {
+    if (!(operandAType.isSignedInteger(8) ||
+          operandAType.isUnsignedInteger(8)) &&
+        (operandBType.isSignedInteger(8) ||
+         operandBType.isUnsignedInteger(8))) {
+      return false;
+    }
+  }
+  return true;
+}
+
+LogicalResult BrgemmOp::verify() {
+  BrgemmOp &brgemmOp = *this;
+
+  SmallVector<Value> inputs = brgemmOp.getInputs();
+  // inputs for BRGEMM: kernel id, A memref, B memref, C memref, batch_size,
+  // addr_len
+  if (inputs.size() != 6) {
+    return brgemmOp.emitOpError() << "expect 6"
+                                  << " inputs but got " << inputs.size();
+  }
+  // Verify the dispatch to be an i64.
+  Value dispatch = brgemmOp.getDispatch();
+  if (!dispatch.getType().isInteger(64)) {
+    return brgemmOp.emitOpError()
+           << "expect an i64 but got " << dispatch.getType()
+           << " for operand 0 (dispatch)";
+  }
+
+  // Verify the compatibility of memref types
+  SmallVector<Value> memrefOperands = {
+      brgemmOp.getOperandA(), brgemmOp.getOperandB(), brgemmOp.getOutput()};
+  SmallVector<Type> typeOperands = {
+      getElementTypeOrSelf(memrefOperands[0].getType()),
+      getElementTypeOrSelf(memrefOperands[1].getType()),
+      getElementTypeOrSelf(memrefOperands[2].getType())};
+  if (!isTypeCompatible(typeOperands[2], typeOperands[0], typeOperands[1])) {
+    return brgemmOp.emitOpError()
+           << "operands types: " << typeOperands[0] << " X " << typeOperands[1]
+           << " -> " << typeOperands[2] << " are imcompatible";
+  }
+
+  // Verify the rank of the shaped operands.
+  for (size_t idx = 0; idx < memrefOperands.size(); idx++) {
```

Comment:
Why is this even a loop? There are checks for each iteration and a remainder

Reply:
Remove redundant loop and change to plain checking codes.